File: 01-week-1--starter-algorithms/00-day-1--reverse-a-string/solutions/reverse_string.rb
def reverse_string(str)
  reversed_str = ""

  str.chars.each do |char|
    reversed_str = char + reversed_str
  end

  reversed_str
end

puts "Expecting: 'ih'"
puts reverse_string('hi')

puts

puts "Expecting: 'ybabtac'"
puts reverse_string('catbaby')

puts

puts "Expecting: '' (empty string)"
puts reverse_string('')

puts

puts "Expecting: 'a'"
puts reverse_string('a')

############################################################################
# Please add your pseudocode to this file:

# initialize reversed_str to store empty string

# iterate over each letter in the input string:
#   place character before previous characters and store result in reversed_str

# return reversed_str
############################################################################

############################################################################
# And a written explanation of your solution:

# The simplest way I can think of reversing a string is to start by thinking of a really
# simple example: "ab"

# To solve for "ab", you take "a", store it in a variable, and then take the next character "b",
# and place it before "a" in that same variable. If we do this for each character, we should be
# able to reverse the string.
############################################################################

File: 01-week-1--starter-algorithms/00-day-1--reverse-a-string/solutions/reverse_string.js
function reverseString(str) {
  let reversed = "";

  for (let i = str.length - 1; i > -1; --i) {
    reversed = reversed + str[i];
  }

  return reversed;
}

console.log("Expecting: 'ih'");
console.log(reverseString('hi'));

console.log("");

console.log("Expecting: 'ybabtac'");
console.log(reverseString('catbaby'));

console.log("");

console.log("Expecting: 'a'");
console.log(reverseString('a'));

console.log("");

console.log("Expecting: '' (empty string)");
console.log(reverseString(''));

// Please add your pseudocode to this file
/**************************************************
 * initialize a variable called reversed with an empty string
 * 
 * iterate backwards through the input string:
 *    set reversed to reversed + current character
 * 
 * return reversed
 * ************************************************/


// And a written explanation of your solution
/**************************************************
 * If I iterate over the input string backwards, I should be able
 * to add whichever character I'm currently iterating over onto
 * my result string (""). For example, if the string is "ab", I'll iterate
 * over "b" first, add that to my result ("b"), and then iterate over
 * "a", and add that to my result next ("ba").
 * ************************************************/
